packages/core/src/sync-realtime/service.ts

```diff
@@ -8,878 +8,633 @@ import {
-    for (let i = blocks.length - 1; i > 1; i--) {
-      if (blocks[i].parentHash !== blocks[i - 1].hash) return false;
+    // Find any logs matching log filters or child contract filters.
+    // NOTE: It might make sense to just insert all logs rather than introduce
+    // a potentially slow DB operation here. It's a tradeoff between sync
+    // latency and database growth.
+    // NOTE: Also makes sense to hold factoryChildAddresses in memory rather than
+    // a query each interval.
+    const factoryLogFilters = await Promise.all(
+      realtimeSyncService.factorySources.map(async (factory) => {
+        const iterator = realtimeSyncService.syncStore.getFactoryChildAddresses(
+          {
+            chainId: realtimeSyncService.network.chainId,
+            factory: factory.criteria,
+            upToBlockNumber,
+          },
+        );
+        const childContractAddresses: Address[] = [];
+        for await (const batch of iterator) {
+          childContractAddresses.push(...batch);
+        }
+        return {
+          address: childContractAddresses,
+          topics: factory.criteria.topics,
+        };
+      }),
+    );
+
+    return filterLogs({
+      logs,
+      logFilters: [
+        ...realtimeSyncService.logFilterSources.map((l) => l.criteria),
+        ...factoryLogFilters,
+      ],
+    });
+  }
+};
+
+export const validateLocalBlockchainState = async (
+  realtimeSyncService: Service,
+  pendingFinalizedBlock: LightBlock,
+) => {
+  realtimeSyncService.common.logger.debug({
+    service: "realtime",
+    msg: `Validating local chain from block ${
+      realtimeSyncService.finalizedBlock.number + 1
+    } to ${pendingFinalizedBlock.number} (network=${
+      realtimeSyncService.network.name
+    })`,
+  });
+
+  const logs = await _eth_getLogs(realtimeSyncService, {
+    fromBlock: realtimeSyncService.finalizedBlock.number + 1,
+    toBlock: pendingFinalizedBlock.number,
+  }).then(sortLogs);
+
+  const remoteLogs = await getMatchedLogs(realtimeSyncService, {
+    logs,
+    insertChildAddressLogs: false,
+    upToBlockNumber: BigInt(pendingFinalizedBlock.number),
+  });
+
+  const localLogs = sortLogs(
+    realtimeSyncService.localChain
+      .filter(({ block }) => block.number <= pendingFinalizedBlock.number)
+      .flatMap(({ logs }) => logs),
+  );
+
+  for (let i = 0; i < localLogs.length && i < remoteLogs.length; i++) {
+    if (
+      remoteLogs[i].blockHash !== localLogs[i].blockHash ||
+      remoteLogs[i].logIndex !== localLogs[i].logIndex
+    ) {
+      return true;
     }
+  }
+
+  if (localLogs.length !== remoteLogs.length) {
     return true;
-  };
+  }
 
-  private getLatestLocalBlock = (): LightBlock =>
-    this.blocks[this.blocks.length - 1] ?? this.finalizedBlock;
-}
+  return false;
+};
+
+const getLatestLocalBlock = ({
+  localChain,
+  finalizedBlock,
+}: Pick<Service, "localChain" | "finalizedBlock">) => {
+  if (localChain.length === 0) {
+    return finalizedBlock;
```

Comment:
Is there a scenario where this would come into play?

Reply:
Immediately upon startup, when the local chain is empty or if the entire local chain was re-orged but the finalized block stayed the same.

Reply:
Makes sense, thanks